src/libstd/thread/local.rs

```diff
@@ -424,36 +529,30 @@ pub mod fast {
 
     unsafe extern fn destroy_value<T>(ptr: *mut u8) {
         let ptr = ptr as *mut Key<T>;
-        // Right before we run the user destructor be sure to flag the
-        // destructor as running for this thread so calls to `get` will return
-        // `None`.
-        (*ptr).dtor_running.set(true);
-
-        // Some implementations may require us to move the value before we drop
-        // it as it could get re-initialized in-place during destruction.
-        //
-        // Hence, we use `ptr::read` on those platforms (to move to a "safe"
-        // location) instead of drop_in_place.
-        if requires_move_before_drop() {
-            ptr::read((*ptr).inner.get());
-        } else {
-            ptr::drop_in_place((*ptr).inner.get());
-        }
+        let tmp = ptr::read((*ptr).inner.get());
+        // Set inner to Destroyed before we drop tmp so that a
+        // recursive call to get (called from the destructor) will
+        // be able to detect that the value is already being dropped.
+        ptr::write((*ptr).inner.get(), __LocalKeyValue::Destroyed);
```

Comment:
I think `mem::replace` can be used here instead of read/write?

Reply:
Good call.